File: src/lib/vault.ts
/**
 * The sync folder — "keep my Verba in iCloud Drive and pick it up on the laptop".
 *
 * ponytail: the SQLite file is **not** moved into that folder, though the plumbing
 * would allow it (tauri-plugin-sql joins an absolute path straight through). A
 * live SQLite database on iCloud Drive or Google Drive is a known way to lose one:
 * the daemon syncs `.db`, `-wal` and `-shm` independently and out of order, iCloud
 * evicts files it thinks are cold, and two machines with the app open at once
 * produce "verba 2.db" instead of a merge. The failure mode is not a bad sync, it
 * is `database disk image is malformed` — a year of history, gone, with no
 * warning and no earlier good copy.
 *
 * So the database stays local, where it is fast and safe, and what lives in the
 * folder is a **snapshot**: one JSON file, written whole, replaced atomically.
 * That is a file a sync daemon is actually good at. The cost is honest — sync is
 * session-granular rather than keystroke-granular, and a machine that dies before
 * its push loses the last few seconds — and for a single learner moving between
 * two machines, that is the right trade.
 *
 * The snapshot is the same envelope the manual export writes (lib/backup.ts), so
 * a folder is just a backup that keeps itself up to date, and either can be fed
 * to the other.
 */

import { parseBackup, restore, collect, deviceId, type Backup } from "./backup.ts";

/** Reached lazily so this module's pure half stays runnable outside Tauri (vault.check.ts). */
const rust = async () => (await import("@tauri-apps/api/core")).invoke;

const DIR_KEY = "verba.vault";
const STATE_KEY = "verba.vaultState";

/** The header beside the data, small enough to read on every launch. */
export interface Meta {
  app: "verba";
  format: number;
  /** When the folder was last written. This is the whole sync clock. */
  updatedAt: number;
  device: string;
  appVersion: string;
}

/**
 * What this machine believes about the folder.
 *
 * `syncedAt` is the folder's `updatedAt` as of the last time the two agreed —
 * set by a push (to what we just wrote) or a pull (to what we just took). It is
 * deliberately the remote's clock and not a local timestamp: comparing two
 * machines' wall clocks across a timezone change or a dead CMOS battery is how
 * sync bugs are born, and this way only one clock is ever read.
 */
export interface State {
  syncedAt: number;
  /** Local changes since `syncedAt` that the folder has not been told about. */
  dirty: boolean;
}

const NO_STATE: State = { syncedAt: 0, dirty: false };

export function vaultDir(): string {
  return localStorage.getItem(DIR_KEY) ?? "";
}

export function state(): State {
  try {
    return { ...NO_STATE, ...JSON.parse(localStorage.getItem(STATE_KEY) ?? "{}") };
  } catch {
    return { ...NO_STATE };
  }
}

function setState(s: State): void {
  localStorage.setItem(STATE_KEY, JSON.stringify(s));
}

/** Something was written locally. Cheap enough to call from every DB write. */
export function markDirty(): void {
  if (!vaultDir()) return;
  const s = state();
  if (!s.dirty) setState({ ...s, dirty: true });
  schedulePush();
}

// ---- the decision ----

export type Plan =
  /** Local is ahead — publish it. */
  | "push"
  /** The folder is ahead — take it. */
  | "pull"
  /** Both moved since they last agreed. Only the learner can choose. */
  | "conflict"
  /** Nothing to do. */
  | "idle";

/**
 * The whole sync policy, as one pure function.
 *
 * It reads two facts: has the folder changed since we last agreed with it
 * (`remote.updatedAt !== syncedAt`), and have we changed since then (`dirty`).
 * Four combinations, four answers, no clocks compared and no heuristics — which
 * is the point. Everything that could go subtly wrong about this feature is
 * decided here, in eight lines that a check file can exhaust.
 */
export function decide(s: State, remote: Meta | null): Plan {
  if (!remote) return "push"; // an empty folder is always ours to fill
  const remoteMoved = remote.updatedAt !== s.syncedAt;
  if (remoteMoved && s.dirty) return "conflict";
  if (remoteMoved) return "pull";
  return s.dirty ? "push" : "idle";
}

/** `parseBackup`'s counterpart for the header — a folder holding somebody else's JSON must not read as a vault. */
export function parseMeta(text: string | null): Meta | null {
  if (!text) return null;
  try {
    const m = JSON.parse(text);
    if (m?.app !== "verba" || typeof m.updatedAt !== "number") return null;
    return {
      app: "verba",
      format: typeof m.format === "number" ? m.format : 1,
      updatedAt: m.updatedAt,
      device: typeof m.device === "string" ? m.device : "another machine",
      appVersion: typeof m.appVersion === "string" ? m.appVersion : "unknown",
    };
  } catch {
    return null;
  }
}

// ---- talking to the folder ----

/** The header, which is all any decision needs. Cheap enough for every launch and every push. */
async function remoteMeta(dir: string): Promise<Meta | null> {
  return parseMeta(await (await rust())<string | null>("vault_meta", { dir }));
}

/** Header *and* data — only when something is about to be restored or described. */
async function load(dir: string): Promise<{ meta: Meta | null; backup: Backup | null }> {
  const meta = await remoteMeta(dir);
  const data = await (await rust())<string | null>("vault_data", { dir });
  // A header with no data beside it is a folder that was copied half-way. Read as
  // empty: pushing over it is right, pulling from it would restore nothing.
  if (!meta || !data) return { meta: null, backup: null };
  try {
    return { meta, backup: parseBackup(data) };
  } catch {
    return { meta, backup: null };
  }
}

/** Hand this machine's whole state to the folder. */
export async function push(appVersion: string): Promise<Meta> {
  const dir = vaultDir();
  const backup = await collect(appVersion);
  const meta: Meta = {
    app: "verba",
    format: backup.format,
    updatedAt: backup.exportedAt,
    device: deviceId(),
    appVersion,
  };
  await (await rust())("vault_save", { dir, meta: JSON.stringify(meta), data: JSON.stringify(backup) });
  setState({ syncedAt: meta.updatedAt, dirty: false });
  return meta;
}

/**
 * Become what the folder holds. The caller reloads the window afterwards —
 * settings, packs and every screen's state were just replaced underneath it,
 * and re-reading them one hook at a time is not something to get right by hand.
 */
export async function pull(): Promise<void> {
  const { meta, backup } = await load(vaultDir());
  if (!meta || !backup) throw new Error("That folder has no Verba data in it yet.");
  await restore(backup);
  setState({ syncedAt: meta.updatedAt, dirty: false });
}

/** Park a copy of what is about to be overwritten. Returns the file it wrote. */
async function park(dir: string, data: string, label: string): Promise<string> {
  return (await rust())<string>("vault_conflict", { dir, data, label });
}

export interface Conflict {
  /** Who wrote the folder's version, and when. */
  remote: Meta;
}

export interface SyncResult {
  plan: Plan;
  /** Set when `plan` is "conflict" — the app has to ask before anything is touched. */
  conflict?: Conflict;
}

/**
 * One round of sync. Safe to call on launch, on a timer, and from a button.
 *
 * A conflict is *reported*, never resolved here. Two machines' worth of work is
 * not something a background task gets to pick between.
 */
export async function sync(appVersion: string): Promise<SyncResult> {
  const dir = vaultDir();
  if (!dir) return { plan: "idle" };

  const meta = await remoteMeta(dir);
  const plan = decide(state(), meta);
  if (plan === "push") await push(appVersion);
  if (plan === "pull") await pull();
  if (plan === "conflict") {
    const conflict = { remote: meta! };
    // Reported *and* announced: a clash found by the Sync-now button deserves the
    // same dialog as one found at launch, and the dialog belongs to the app shell
    // rather than to whichever screen happened to ask.
    onConflict?.(conflict);
    return { plan, conflict };
  }
  return { plan };
}

/**
 * The learner's answer to a conflict. Whichever side loses is written to the
 * folder as `verba-conflict-<device>-<date>.json` first — the losing side is
 * still a real day of someone's work, and it stays importable by hand.
 */
export async function resolve(choice: "mine" | "theirs", appVersion: string): Promise<void> {
  const dir = vaultDir();
  const stamp = new Date().toISOString().slice(0, 16).replace(":", "");

  if (choice === "mine") {
    const { meta, backup } = await load(dir);
    if (backup) await park(dir, JSON.stringify(backup), `${meta?.device ?? "folder"}-${stamp}`);
    await push(appVersion);
    return;
  }
  await park(dir, JSON.stringify(await collect(appVersion)), `${deviceId()}-${stamp}`);
  await pull();
}

/**
 * Point at a folder, and report what is already there so the caller can ask the
 * only question that matters — "there is data on both sides, which one wins?".
 *
 * Nothing is written and nothing is restored here. Attaching to a folder is not
 * itself a sync; the app is left in a state where the next `sync()` would pull,
 * which is what a learner picking up an existing folder wants, and what the
 * caller overrides by pushing if they chose the other way.
 */
export async function attach(dir: string): Promise<{ meta: Meta | null; summary: Backup | null }> {
  await (await rust())("vault_check", { dir });
  const { meta, backup } = await load(dir);
  localStorage.setItem(DIR_KEY, dir);
  setState({ syncedAt: 0, dirty: true }); // both sides unproven until the caller says which wins
  return { meta, summary: backup };
}

/** The folder picker, here rather than in a view so setup and Settings open the same one. */
export async function pickFolder(): Promise<string | null> {
  const { open } = await import("@tauri-apps/plugin-dialog");
  const picked = await open({ directory: true, multiple: false });
  return typeof picked === "string" ? picked : null;
}

/** Stop syncing. The folder keeps its file — detaching is not a delete. */
export function detach(): void {
  localStorage.removeItem(DIR_KEY);
  localStorage.removeItem(STATE_KEY);
}

// ---- when a push happens ----

/**
 * Writes are bursty — a finished conversation is a session row, a dozen messages,
 * a metrics row and a handful of cards inside a second — and each push serialises
 * the entire database. Coalescing them costs a few seconds of staleness and turns
 * twenty snapshots into one.
 */
const PUSH_DELAY_MS = 4000;

let timer: ReturnType<typeof setTimeout> | null = null;
let version = "0.0.0";
let onError: ((e: unknown) => void) | null = null;
let onConflict: ((c: Conflict) => void) | null = null;

/** Wire the module to the running app: which version stamps a snapshot, and where a failure or a clash surfaces. */
export function configure(
  appVersion: string,
  handlers: { error: (e: unknown) => void; conflict: (c: Conflict) => void },
): void {
  version = appVersion;
  onError = handlers.error;
  onConflict = handlers.conflict;
}

function schedulePush(): void {
  if (timer) clearTimeout(timer);
  timer = setTimeout(() => {
    timer = null;
    void flush();
  }, PUSH_DELAY_MS);
}

/**
 * Push now if there is anything to push.
 *
 * Called on the debounce, and again when the window is hidden — closing the lid
 * on a finished session should not cost the session.
 *
 * It re-reads the folder's header first, and that check is not ceremony: the app
 * can sit open for hours, and the other machine can publish at any point during
 * them. Without it, a push here would be a blind overwrite of work that arrived
 * after launch — the one way this feature could actually destroy something. It
 * costs one small file read per push, and it routes the clash to the same
 * conflict dialog a launch would.
 */
export async function flush(): Promise<void> {
  const dir = vaultDir();
  if (!dir || !state().dirty) return;
  try {
    const meta = await remoteMeta(dir);
    if (decide(state(), meta) === "conflict") return onConflict?.({ remote: meta! });
    await push(version);
  } catch (e) {
    onError?.(e);
  }
}

File: src-tauri/src/vault.rs
// Reading and writing the learner's data *outside* the app's own directory —
// a backup file they chose, or a folder they sync (iCloud Drive, Google Drive,
// Syncthing, a USB stick).
//
// ponytail: this is Rust rather than tauri-plugin-fs on purpose. The whole point
// is an arbitrary path the learner names at runtime, and the fs plugin's answer
// to that is either a permissive scope in capabilities/ — which hands *every*
// path to the webview forever — or a runtime scope call that has to be threaded
// through anyway. Four commands here keep the webview's reach to exactly the
// files this feature is about.
//
// Nothing in here knows what a backup contains. It moves strings.

use std::fs;
use std::path::{Path, PathBuf};

/// The two files a synced folder holds. `data` is the whole backup envelope
/// (see src/lib/backup.ts); `meta` is the small header — who wrote it, when,
/// which app version — that a sync decision can be made from without parsing
/// and reading megabytes of passages.
pub const DATA_FILE: &str = "verba-data.json";
pub const META_FILE: &str = "verba-meta.json";

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("{0}")]
    Msg(String),
    #[error("{0}")]
    Io(#[from] std::io::Error),
}

impl serde::Serialize for Error {
    fn serialize<S: serde::ser::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
        s.serialize_str(&self.to_string())
    }
}

/// Write a file so that a reader never sees a half-written one.
///
/// This matters more here than anywhere else in the app: the destination is a
/// folder a sync daemon is *watching*, and a daemon that uploads the file while
/// we are still writing it would publish a truncated backup to every other
/// machine. Write a sibling temp file, flush it, then rename — rename within a
/// directory is atomic on every filesystem we ship to, so the daemon sees the
/// old file or the new one and never something in between.
fn write_atomic(path: &Path, contents: &str) -> Result<(), Error> {
    let dir = path
        .parent()
        .ok_or_else(|| Error::Msg(format!("{} has no parent directory", path.display())))?;
    fs::create_dir_all(dir)?;

    // Named for the target so two concurrent writes to different files in the
    // same folder cannot collide on the temp name.
    let tmp = dir.join(format!(
        ".{}.tmp",
        path.file_name().and_then(|n| n.to_str()).unwrap_or("verba")
    ));
    fs::write(&tmp, contents)?;
    fs::rename(&tmp, path)?;
    Ok(())
}

/// `None` for "no such file", an error for a file that exists and cannot be
/// read. The caller needs to tell those apart: the first is an empty folder
/// waiting for its first sync, the second is a real problem worth showing.
fn read_opt(path: &Path) -> Result<Option<String>, Error> {
    match fs::read_to_string(path) {
        Ok(s) => Ok(Some(s)),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(None),
        Err(e) => Err(e.into()),
    }
}

/// The header alone.
///
/// Split from the data on purpose: every sync decision is made from `updatedAt`
/// and nothing else, and this is the call the app makes on launch, before each
/// push, and whenever a window comes back. Reading it must not mean pulling a
/// year of passages across the IPC bridge — the header is a few hundred bytes,
/// the data can be tens of megabytes.
#[tauri::command]
pub fn vault_meta(dir: String) -> Result<Option<String>, Error> {
    read_opt(&PathBuf::from(dir).join(META_FILE))
}

/// The backup itself. Only read when something is actually being restored.
#[tauri::command]
pub fn vault_data(dir: String) -> Result<Option<String>, Error> {
    read_opt(&PathBuf::from(dir).join(DATA_FILE))
}

/// Publish a backup to the folder.
///
/// Data first, header second, and never the other way round: the header is what
/// another machine reads to decide "this folder is newer than me, pull it". If
/// it landed first and the write died — the disk filled, the laptop slept, the
/// sync daemon choked — that machine would be told to pull a backup that is
/// still the previous one, and would silently roll itself back. Written in this
/// order, a failure between the two leaves a folder whose header is *older*
/// than its data, which costs one redundant push and loses nothing.
#[tauri::command]
pub fn vault_save(dir: String, meta: String, data: String) -> Result<(), Error> {
    let dir = PathBuf::from(dir);
    write_atomic(&dir.join(DATA_FILE), &data)?;
    write_atomic(&dir.join(META_FILE), &meta)?;
    Ok(())
}

/// Park a copy that would otherwise be overwritten, and return where it went.
///
/// Used when two machines both wrote since they last agreed. Whichever side
/// loses, its data still exists as a file the learner can import by hand — this
/// feature is not allowed to be the reason a year of history disappears.
#[tauri::command]
pub fn vault_conflict(dir: String, data: String, label: String) -> Result<String, Error> {
    let safe: String = label
        .chars()
        .map(|c| if c.is_alphanumeric() || c == '-' { c } else { '-' })
        .collect();
    let path = PathBuf::from(dir).join(format!("verba-conflict-{safe}.json"));
    write_atomic(&path, &data)?;
    Ok(path.display().to_string())
}

/// Prove the folder can actually be written to, at the moment the learner picks
/// it rather than at the first sync an hour later. A read-only volume, a
/// not-yet-downloaded iCloud directory and a typo'd path all fail here, where
/// the error still has a folder picker next to it to explain itself.
#[tauri::command]
pub fn vault_check(dir: String) -> Result<(), Error> {
    let dir = PathBuf::from(dir);
    if !dir.is_dir() {
        return Err(Error::Msg(format!("{} is not a folder", dir.display())));
    }
    let probe = dir.join(".verba-write-test");
    fs::write(&probe, b"ok")?;
    fs::remove_file(&probe)?;
    Ok(())
}

/// One file, one path — the manual "Export my data" half, where the learner
/// named the exact file in a save dialog.
#[tauri::command]
pub fn file_write(path: String, contents: String) -> Result<(), Error> {
    write_atomic(Path::new(&path), &contents)
}

#[tauri::command]
pub fn file_read(path: String) -> Result<String, Error> {
    Ok(fs::read_to_string(path)?)
}
